server/etcdserver/server.go

```diff
@@ -2170,6 +2183,18 @@ func (s *EtcdServer) snapshot(snapi uint64, confState raftpb.ConfState) {
 		"saved snapshot",
 		zap.Uint64("snapshot-index", snap.Metadata.Index),
 	)
+}
+
+func (s *EtcdServer) maybeCompactRaftLog(ep *etcdProgress) {
+	lg := s.Logger()
+
+	// Keep some in memory log entries for slow followers, while keeping the entries up to snapshot index.
+	// Only compact raft log once every N applies
+	if ep.appliedi <= ep.snapi+s.Cfg.SnapshotCatchUpEntries || ep.appliedi%s.Cfg.CompactRaftLogEveryNApplies != 0 {
+		return
+	}
+
+	compacti := ep.appliedi - s.Cfg.SnapshotCatchUpEntries
```

Comment:
```suggestion
	// Retain all log entries up to the latest snapshot index to ensure any member can recover from that snapshot. 
    // Beyond the snapshot index, preserve the most recent s.Cfg.SnapshotCatchUpEntries entries in memory. 
    // This allows slow followers to catch up by synchronizing entries instead of requiring a full snapshot transfer.
	// Only compact raft log once every N applies
	if ep.snapi > s.Cfg.SnapshotCatchUpEntries  {
		return
	}

	compacti := ep.snapi - s.Cfg.SnapshotCatchUpEntries
```

In the first PR, let's keep the same number of entries, just make the compaction independent from snapshot.